examples/src/main/python/ml/bucketed_random_projection_lsh_example.py

```diff
@@ -0,0 +1,86 @@
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from __future__ import print_function
+
+# $example on$
+from pyspark.ml.feature import BucketedRandomProjectionLSH
+from pyspark.ml.linalg import Vectors
+# $example off$
+from pyspark.sql import SparkSession
+
+"""
+An example demonstrating BucketedRandomProjectionLSH.
+Run with:
+  bin/spark-submit examples/src/main/python/ml/bucketed_random_projection_lsh_example.py
+"""
+
+if __name__ == "__main__":
+    spark = SparkSession \
+        .builder \
+        .appName("BucketedRandomProjectionLSHExample") \
+        .getOrCreate()
+
+    # $example on$
+    dataA = [(0, Vectors.dense([1.0, 1.0]),),
+             (1, Vectors.dense([1.0, -1.0]),),
+             (2, Vectors.dense([-1.0, -1.0]),),
+             (3, Vectors.dense([-1.0, 1.0]),)]
+    dfA = spark.createDataFrame(dataA, ["id", "features"])
+
+    dataB = [(4, Vectors.dense([1.0, 0.0]),),
+             (5, Vectors.dense([-1.0, 0.0]),),
+             (6, Vectors.dense([0.0, 1.0]),),
+             (7, Vectors.dense([0.0, -1.0]),)]
+    dfB = spark.createDataFrame(dataB, ["id", "features"])
+
+    key = Vectors.dense([1.0, 0.0])
+
+    brp = BucketedRandomProjectionLSH(inputCol="features", outputCol="hashes", bucketLength=2.0,
+                                      numHashTables=3)
+    model = brp.fit(dfA)
+
+    # Feature Transformation
+    print("The hashed dataset where hashed values are stored in the column 'values':")
+    model.transform(dfA).show()
+    # Cache the transformed columns
+    transformedA = model.transform(dfA).cache()
+    transformedB = model.transform(dfB).cache()
+
+    # Approximate similarity join
+    print("Approximately joining dfA and dfB on distance smaller than 1.5:")
+    model.approxSimilarityJoin(dfA, dfB, 1.5)\
+        .select("datasetA.id", "datasetB.id", "distCol").show()
+    print("Joining cached datasets to avoid recomputing the hash values:")
+    model.approxSimilarityJoin(transformedA, transformedB, 1.5)\
+        .select("datasetA.id", "datasetB.id", "distCol").show()
+
+    # Self Join
+    print("Approximately self join of dfB on distance smaller than 2.5:")
+    model.approxSimilarityJoin(dfA, dfA, 2.5).filter("datasetA.id < datasetB.id")\
+        .select("datasetA.id", "datasetB.id", "distCol").show()
+
+    # Approximate nearest neighbor search
+    print("Approximately searching dfA for 2 nearest neighbors of the key:")
+    model.approxNearestNeighbors(dfA, key, 2).show()
+    print("Searching cached dataset to avoid recomputing the hash values:")
```

Comment:
Still think this is confusing here too. maybe we could leave comments:

````python
# compute the locality-sensitive hashes for the input rows, then perform approximate nearest neighbor search.
# we could avoid computing hashes by passing in the already-transformed dataframe, e.g. 
# `model.approxNearestNeighbors(transformedA, key, 2)`
model.approxNearestNeighbors(dfA, key, 2).show()
````

This seems better to me because before you'd have to read the code to see the difference between the two cases anyway. Now, those who read the code can _still_ see the difference from the comment, and those who just run the example from the command line will not be confused or waste time trying to understand why we output the exact same table twice with different explanations.